src/context.rs

```diff
@@ -0,0 +1,354 @@
+        }
+        let data_container = self
+            .plugin_data
+            .get(type_id)
+            .unwrap()
+            .downcast_ref::<T::DataContainer>();
+        match data_container {
+            Some(x) => Some(x),
+            None => panic!("Plugin data container of incorrect type"),
+        }
+    }
+
+    pub fn get_time(&self) -> f64 {
+        self.time
+    }
+
+    pub fn add_component<T: Component>(&mut self) {
+        T::init(self);
+    }
+
+    fn add_handlers<E: Copy + 'static>(
+        event_handlers: &mut HashMap<TypeId, Box<dyn Any>>,
+        callback: impl Fn(&mut Context, E) + 'static,
+    ) {
+        let callback_vec = event_handlers
+            .entry(TypeId::of::<E>())
+            .or_insert_with(|| Box::<Vec<Rc<EventHandler<E>>>>::default());
+        let callback_vec: &mut Vec<Rc<EventHandler<E>>> = callback_vec.downcast_mut().unwrap();
+        callback_vec.push(Rc::new(callback));
+    }
+
+    pub fn subscribe_to_event<E: Copy + 'static>(
+        &mut self,
+        callback: impl Fn(&mut Context, E) + 'static,
+    ) {
+        Self::add_handlers(&mut self.event_handlers, callback);
+    }
+
+    pub fn subscribe_immediately_to_event<E: Copy + 'static>(
+        &mut self,
+        callback: impl Fn(&mut Context, E) + 'static,
+    ) {
+        Self::add_handlers(&mut self.immediate_event_handlers, callback);
+    }
+
+    fn collect_callbacks<E: Copy + 'static>(
+        event_handlers: &HashMap<TypeId, Box<dyn Any>>,
+        event: E,
+    ) -> Vec<Box<Callback>> {
+        let mut callbacks_to_return = Vec::<Box<Callback>>::new();
+        let callback_vec = event_handlers.get(&TypeId::of::<E>());
+        if let Some(callback_vec) = callback_vec {
+            let callback_vec: &Vec<Rc<EventHandler<E>>> = callback_vec.downcast_ref().unwrap();
+            if !callback_vec.is_empty() {
+                for callback in callback_vec {
+                    let internal_callback = Rc::clone(callback);
+                    callbacks_to_return
+                        .push(Box::new(move |context| internal_callback(context, event)));
+                }
+            }
+        }
+        callbacks_to_return
+    }
+
+    pub fn release_event<E: Copy + 'static>(&mut self, event: E) {
+        // Queue standard handlers
+        for callback in Self::collect_callbacks(&self.event_handlers, event) {
+            self.queue_callback(callback);
+        }
+        // Process immediate handlers
+        for callback in Self::collect_callbacks(&self.immediate_event_handlers, event) {
+            callback(self);
+        }
+    }
+
+    pub fn execute(&mut self) {
+        // Execute callbacks if there are any in the queue
+        loop {
+            let callback = self.callback_queue.pop_front();
+            match callback {
+                Some(callback) => callback(self),
+                None => break,
+            }
+        }
+        // Start plan loop
+        loop {
```

Comment:
I think this could use tests/we could maybe find a way to do the callback draining better, but probably better as a follow up